src/main/drivers/flash_w25n01g.c

```diff
@@ -580,6 +590,209 @@ void w25n01g_pageProgramFinish(flashDevice_t *fdevice)
         programStartAddress = programLoadAddress;
     }
 }
+#else
+void w25n01g_pageProgramBegin(flashDevice_t *fdevice, uint32_t address, void (*callback)(uint32_t length))
+{
+    fdevice->callback = callback;
+    fdevice->currentWriteAddress = address;
+
+}
+
+static uint32_t currentPage = UINT32_MAX;
+
+// Called in ISR context
+// Check if the status was busy and if so repeat the poll
+busStatus_e w25n01g_callbackReady(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+    extDevice_t *dev = fdevice->io.handle.dev;
+
+    uint8_t readyPoll = dev->bus->curSegment->u.buffers.rxData[2];
+
+    if (readyPoll & W25N01G_STATUS_FLAG_BUSY) {
+        return BUS_BUSY;
+    }
+
+    // Bus is now known not to be busy
+    fdevice->couldBeBusy = false;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// A write enable has just been issued
+busStatus_e w25n01g_callbackWriteEnable(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+
+    // As a write has just occurred, the device could be busy
+    fdevice->couldBeBusy = true;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// A write enable has just been issued
+busStatus_e w25n01g_callbackProgExecute(uint32_t arg)
+{
+    UNUSED(arg);
+
+    bufferDirty = false;
+
+    return BUS_READY;
+}
+
+// Called in ISR context
+// Write operation has just completed
+busStatus_e w25n01g_callbackWriteComplete(uint32_t arg)
+{
+    flashDevice_t *fdevice = (flashDevice_t *)arg;
+
+    fdevice->currentWriteAddress += fdevice->callbackArg;
+
+    // Call transfer completion callback
+    if (fdevice->callback) {
+        fdevice->callback(fdevice->callbackArg);
+    }
+
+    return BUS_READY;
+}
+
+uint32_t w25n01g_pageProgramContinue(flashDevice_t *fdevice, uint8_t const **buffers, uint32_t *bufferSizes, uint32_t bufferCount)
+{
+    if (bufferCount < 1) {
+        fdevice->callback(0);
+        return 0;
+    }
+
+    // The segment list cannot be in automatic storage as this routine is non-blocking
+    STATIC_DMA_DATA_AUTO uint8_t readStatus[] = { W25N01G_INSTRUCTION_READ_STATUS_REG, W25N01G_STAT_REG, 0 };
+    STATIC_DMA_DATA_AUTO uint8_t readyStatus[3];
+    STATIC_DMA_DATA_AUTO uint8_t writeEnable[] = { W25N01G_INSTRUCTION_WRITE_ENABLE };
+    STATIC_DMA_DATA_AUTO uint8_t progExecCmd[] = { W25N01G_INSTRUCTION_PROGRAM_EXECUTE, 0, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progExecCmd2[] = { W25N01G_INSTRUCTION_PROGRAM_EXECUTE, 0, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progExecDataLoad[] = { W25N01G_INSTRUCTION_PROGRAM_DATA_LOAD, 0, 0};
+    STATIC_DMA_DATA_AUTO uint8_t progRandomProgDataLoad[] = { W25N01G_INSTRUCTION_RANDOM_PROGRAM_DATA_LOAD, 0, 0};
```

Comment:
Maybe this may be more readable:
```suggestion
    STATIC_DMA_DATA_AUTO struct { uint8_t cmd, uint8_t addr[2] } progRandomProgDataLoad[] = { W25N01G_INSTRUCTION_RANDOM_PROGRAM_DATA_LOAD};
```
But because structure alignment, only uint8_t is safe inside struct. 

And then:
```
progRandomProgDataLoad.addr[0] = (columnAddress >> 8) & 0xff;

Reply:
I deliberately coded this as a byte sequence which can be easily reconciled against the datasheet definition for the device.

Reply:
Ok, but it is very hard to read without datasheet. Maybe add comment that will explain command parameters?